pkg/limits/frontend/ring.go

```diff
@@ -62,10 +64,32 @@ func (g *RingStreamUsageGatherer) forGivenReplicaSet(ctx context.Context, rs rin
 	}
 
 	errg, ctx := errgroup.WithContext(ctx)
-	responses := make([]GetStreamUsageResponse, len(rs.Instances))
+	var owningInstances []ring.InstanceDesc
+
+outer:
+	for _, hash := range r.StreamHashes {
+		for _, instance := range rs.Instances {
```

Comment:
I don't really understand why you've chose three nested loops here (including lines 78 to 82) to calculate which instance owns which stream? If we have 20 stream hashes, 50 instances, and 50 owning instances, we are looking at almost 50,000 iterations in the worst case?

Reply:
What I would recommend here is inverting the data structure we return from `getPartitionConsumers`, you can then do it in linear time.